### ISSUE TITLE
Application is crashing even if payment is successful

### ISSUE THREAD

**Issue by yashwp** [mdazharuddin1011999/UPI-Plugin-Flutter#21]
I'm using your the example given in the README, with some very minor tweaks on the UI level only. I'm sending money to an account using PayTm. The payment is successful, but when it should return back to the application, the application itself is **crashing** with the below error:-

```
D/AndroidRuntime(10354): Shutting down VM
E/AndroidRuntime(10354): FATAL EXCEPTION: main
E/AndroidRuntime(10354): Process: com.example.stackfinance.stack_app, PID: 10354
E/AndroidRuntime(10354): java.lang.RuntimeException: Failure delivering result ResultInfo{who=null, request=5120, result=-1, data=Intent { (has extras) }} to activity {com.example.stackfinance.stack_app/com.example.stackfinance.stack_app.MainActivity}: java.lang.NullPointerException: Attempt to invoke interface method 'void io.flutter.plugin.common.MethodChannel$Result.success(java.lang.Object)' on a null object reference
E/AndroidRuntime(10354): 	at android.app.ActivityThread.deliverResults(ActivityThread.java:4962)
E/AndroidRuntime(10354): 	at android.app.ActivityThread.handleSendResult(ActivityThread.java:5003)
E/AndroidRuntime(10354): 	at android.app.servertransaction.ActivityResultItem.execute(ActivityResultItem.java:51)
E/AndroidRuntime(10354): 	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
E/AndroidRuntime(10354): 	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
E/AndroidRuntime(10354): 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2109)
E/AndroidRuntime(10354): 	at android.os.Handler.dispatchMessage(Handler.java:107)
E/AndroidRuntime(10354): 	at android.os.Looper.loop(Looper.java:214)
E/AndroidRuntime(10354): 	at android.app.ActivityThread.main(ActivityThread.java:7682)
E/AndroidRuntime(10354): 	at java.lang.reflect.Method.invoke(Native Method)
E/AndroidRuntime(10354): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:516)
E/AndroidRuntime(10354): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:950)
E/AndroidRuntime(10354): Caused by: java.lang.NullPointerException: Attempt to invoke interface method 'void io.flutter.plugin.common.MethodChannel$Result.success(java.lang.Object)' on a null object reference
E/AndroidRuntime(10354): 	at com.cashfree.cashfree_pg.CashfreePgPlugin$1.onActivityResult(CashfreePgPlugin.java:112)
E/AndroidRuntime(10354): 	at io.flutter.embedding.engine.FlutterEnginePluginRegistry$FlutterEngineActivityPluginBinding.onActivityResult(FlutterEnginePluginRegistry.java:691)
E/AndroidRuntime(10354): 	at io.flutter.embedding.engine.FlutterEnginePluginRegistry.onActivityResult(FlutterEnginePluginRegistry.java:378)
E/AndroidRuntime(10354): 	at io.flutter.embedding.android.FlutterActivityAndFragmentDelegate.onActivityResult(FlutterActivityAndFragmentDelegate.java:619)
E/AndroidRuntime(10354): 	at io.flutter.embedding.android.FlutterActivity.onActivityResult(FlutterActivity.java:584)
E/AndroidRuntime(10354): 	at android.app.Activity.dispatchActivityResult(Activity.java:8135)
E/AndroidRuntime(10354): 	at android.app.ActivityThread.deliverResults(ActivityThread.java:4955)
E/AndroidRuntime(10354): 	... 11 more
I/Process (10354): Sending signal. PID: 10354 SIG: 9
```

Here is my code - 
```
class UpiScreen extends StatelessWidget {
  var _upiService = UpiService();
  var _txnFuture;
  List<UpiApp> list;

  UpiScreen() {
    _invokeUpi();
  }

  _invokeUpi() async {
    list = await _upiService.init();
  }

  Future<UpiResponse> _invokeApp(app) async {
    return await _upiService.initiateTransaction(
        amount: 1.0,
        app: UpiApp.PayTM,
        receiverName: 'Name Lastname',
        receiverUpiId: '****.***@okhdfcbank',
        txnNote: 'TESTING');
  }

  Widget _displayUpiApps() {
    if (list == null)
      return Center(child: CircularProgressIndicator());
    else if (list.length == 0)
      return Center(child: Text("No list found to handle transaction."));
    else
      return ListView(
        children: list.map<Widget>((UpiApp e) {
          return ListTile(
            onTap: () {
              _txnFuture = _invokeApp(e.app);
            },
            leading: Image.memory(
              e.icon,
              height: 60,
              width: 60,
            ),
            title: Text(e.name),
          );
        }).toList(),
      );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('UPI'),
      ),
      body: Column(
        children: <Widget>[
          Container(height: 200, child: _displayUpiApps()),
          if (_txnFuture != null)
            Expanded(
              child: FutureBuilder(
                future: _txnFuture,
                builder: (context,  AsyncSnapshot<UpiResponse> snapshot) {
                  if (snapshot.connectionState == ConnectionState.waiting) {
                    return Loader(color: Variables.cashColor);
                  } else {
                    print('Show the error - ${snapshot.error}'); // These print statement never called.

                    if (snapshot.error != null) {
                      return Text('An Unknown error has occured');
                    }

                    UpiResponse _upiResponse;
                    _upiResponse = snapshot.data;
                    print('UPI response $_upiResponse');

                    if (_upiResponse.error != null) {
                      String text = '';
                      _upiService.errorHandler(snapshot.data);  // See the UpiService class below

                      return  Text(text);
                    }

                    String txnId = _upiResponse.transactionId;
                    String resCode = _upiResponse.responseCode;
                    String txnRef = _upiResponse.transactionRefId;
                    String status = _upiResponse.status;
                    String approvalRef = _upiResponse.approvalRefNo;
                    _upiService.checkStatus(status);  // See the UpiService class below

                    return Column(
                      mainAxisAlignment: MainAxisAlignment.center,
                      children: <Widget>[
                        Text('Transaction Id: $txnId\n'),
                        Text('Response Code: $resCode\n'),
                        Text('Reference Id: $txnRef\n'),
                        Text('Status: $status\n'),
                        Text('Approval No: $approvalRef'),
                      ],
                    );
                  }
                },
              ),
            )
        ],
      ),
    );
  }
}
```

### `UpiService`
```
class UpiService {
  UpiIndia _upi = UpiIndia();
  List<UpiApp> _apps = [];

  UpiService() {}

  List<UpiApp> get upiApps {
    return [..._apps];
  }

  Future<List<UpiApp>> init() async {
    _apps = await _upi.getAllUpiApps();
    return [...upiApps];
  }

  Future<UpiResponse> initiateTransaction({
    app,
    receiverUpiId,
    receiverName,
    amount,
    txnNote,
    txtReferenceId,
  }) {
    return _upi.startTransaction(
      app: app,
      receiverUpiId: receiverUpiId,
      receiverName: receiverName,
      amount: amount,
      transactionNote: txnNote,
      transactionRefId: txtReferenceId,
    );
  }

  void checkStatus(status) {
    switch (status) {
      case UpiPaymentStatus.SUCCESS:
        print('Transaction Successful');
        break;
      case UpiPaymentStatus.SUBMITTED:
        print('Transaction Submitted');
        break;
      case UpiPaymentStatus.FAILURE:
        print('Transaction Failed');
        break;
      default:
        print('Received an Unknown transaction status');
    }
  }

  void errorHandler(response) {
    if (response.error != null) {
      switch (response.error) {
        case UpiError.APP_NOT_INSTALLED:
          print('Requested app not installed on device');
          break;
        case UpiError.INVALID_PARAMETERS:
          print('Requested app cannot handle the transaction');
          break;
        case UpiError.NULL_RESPONSE:
          print('Requested app did not returned any response');
          break;
        case UpiError.USER_CANCELLED:
          print('You cancelled the transaction');
          break;
      }
    }
  }
}
```

**Comment by azhar1038** [comment]
Don't know exactly what caused this. I will look into the matter as soon as I am free. If you can tell the exact situation when this happened, it will be useful for me. Also is it happening only in paytm or with others too?

I am a bit busy and may need about a week to check into this. In the mean time if you discover what is causing this problem, feel free to update me or send a pr.

**Comment by yashwp** [comment]
@mdazharuddin1011999 This is happening for GooglePay and Paytm for sure. I don't have other UPI apps like PhonePe and all. But I'm sure they will also crash, as this looks like some internal error.

Steps - 
1. From the list of apps I'm selecting one by tapping 
```
onTap: () {
   _txnFuture = _invokeApp(e.app);
},
```

2. The Paytm app opened with all the required parameters in place.
3. The payment is successfully done.
4. But when it should go back to my app _(the app which invoked UPI)_, it crashed.

**Comment by azhar1038** [comment]
I have changed the implementation to use the newer embedding v2 (with FlutterPlugin class introduced with Flutter 1.12.x). Maybe this will solve the problem. I haven't got time to test it thoroughly yet. If you can pull from the beta branch of repo directly and test if it still crashes, it would be useful for me.

**Comment by azhar1038** [comment]
Are you still seeing this in the beta version? Also I made lots of changes in beta branch today so make sure to go through Changelog and Readme if you want to use this. I will push it to pub after verification.

**Comment by azhar1038** [comment]
[Dev version 2.2.0-dev.1](https://pub.dev/packages/upi_india/versions/2.2.0-dev.1) has been released in pub. You can directly try from there if you want!

**Comment by yashwp** [comment]
@mdazharuddin1011999 I'll check this with the dev version today and let you know if it is working. Thanks for your efforts though.

**Comment by yashwp** [comment]
@mdazharuddin1011999 I have tested with Paytm and GooglePay. It is working like charm. I'll update the example in PR #26.

Let me know when can you publish this updated flow.

**Comment by azhar1038** [comment]
Published version 2.2.0 in pub
